File: Reto4/userMgmtMicroservice/eureka/eureka.go
package eureka

import (
	"bytes"
	"fmt"
	"net/http"
	"os"
	"strings"
	"time"
)

var (
	appName    = getenv("APP_NAME", "user-mgmt-service")
	eurekaCons = strings.TrimRight(getenv("EUREKA_SERVER", "http://localhost:8761/eureka/apps"), "/")
	hostname   = getenv("EUREKA_HOSTNAME", "localhost")
	port       = getenv("SERVER_PORT", "8083")
	instanceId = fmt.Sprintf("%s:%s:%s", hostname, appName, port)
	client     = &http.Client{Timeout: 5 * time.Second}
	ticker     *time.Ticker
)

func RegisterAndHeartbeat() {
	// Intentos de registro con backoff
	for attempt := 0; attempt < 5; attempt++ {
		if register() {
			break
		}
		time.Sleep(time.Duration(1<<attempt) * time.Second)
	}
	interval := 10
	if v := os.Getenv("EUREKA_HEARTBEAT_INTERVAL"); v != "" {
		fmt.Sscanf(v, "%d", &interval)
	}
	ticker = time.NewTicker(time.Duration(interval) * time.Second)
	go func() {
		for range ticker.C {
			heartbeat()
		}
	}()
}

func register() bool {
	url := fmt.Sprintf("%s/%s", eurekaCons, strings.ToUpper(appName))
	payload := fmt.Sprintf(`<instance>
  <instanceId>%s</instanceId>
  <hostName>%s</hostName>
  <app>%s</app>
  <ipAddr>127.0.0.1</ipAddr>
  <status>UP</status>
  <port enabled="true">%s</port>
  <securePort enabled="false">443</securePort>
  <homePageUrl>http://%s:%s/</homePageUrl>
  <statusPageUrl>http://%s:%s/health</statusPageUrl>
  <healthCheckUrl>http://%s:%s/health</healthCheckUrl>
  <vipAddress>%s</vipAddress>
  <dataCenterInfo class="com.netflix.appinfo.InstanceInfo$DefaultDataCenterInfo">
    <name>MyOwn</name>
  </dataCenterInfo>
</instance>`, instanceId, hostname, strings.ToUpper(appName), port, hostname, port, hostname, port, hostname, port, appName)

	req, _ := http.NewRequest("POST", url, bytes.NewBufferString(payload))
	req.Header.Set("Content-Type", "application/xml")
	req.Header.Set("Accept", "application/xml")
	resp, err := client.Do(req)
	if err != nil {
		fmt.Printf("Eureka register error: %v\n", err)
		return false
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusNoContent && resp.StatusCode != http.StatusOK {
		fmt.Printf("Eureka register failed: status=%d\n", resp.StatusCode)
		return false
	}
	fmt.Printf("Registered in eureka: %s\n", url)
	return true
}

func heartbeat() {
	url := fmt.Sprintf("%s/%s/%s", eurekaCons, strings.ToUpper(appName), instanceId)
	req, _ := http.NewRequest("PUT", url, nil)
	resp, err := client.Do(req)
	if err != nil {
		fmt.Printf("Eureka heartbeat error: %v\n", err)
		return
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusNotFound {
		fmt.Println("Eureka heartbeat 404 → re-registering")
		_ = register()
	}
}

func Deregister() {
	if ticker != nil {
		ticker.Stop()
	}
	url := fmt.Sprintf("%s/%s/%s", eurekaCons, strings.ToUpper(appName), instanceId)
	req, _ := http.NewRequest("DELETE", url, nil)
	_, err := client.Do(req)
	if err != nil {
		fmt.Printf("Eureka deregister error: %v\n", err)
	}
	fmt.Println("Deregistered from Eureka.")
}

func getenv(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}

File: Reto4/orderMgmtMicroservice/eureka.py
import os
import asyncio
import requests

# variables de entorno
EUREKA_ENABLED = os.getenv("EUREKA_ENABLED", "1").lower() not in ("0", "false", "no")
EUREKA_SERVER = os.getenv("EUREKA_SERVER", "http://localhost:8761/eureka").rstrip("/")
APP_NAME = os.getenv("APP_NAME", "order-microservice")
FORCED_HOST = (os.getenv("EUREKA_HOSTNAME", "localhost").strip() or "localhost")
PORT = int(os.getenv("PORT", "8082"))
HEARTBEAT_INTERVAL = int(os.getenv("EUREKA_HEARTBEAT_INTERVAL", "10"))


INSTANCE_ID = f"{FORCED_HOST}:{APP_NAME}:{PORT}"

_registered = False
_heartbeat_task: asyncio.Task | None = None


def _build_instance_xml() -> str:
    return f"""<instance>
  <instanceId>{INSTANCE_ID}</instanceId>
  <hostName>{FORCED_HOST}</hostName>
  <app>{APP_NAME}</app>
  <ipAddr>127.0.0.1</ipAddr>
  <status>UP</status>
  <port enabled="true">{PORT}</port>
  <healthCheckUrl>http://localhost:{PORT}/health</healthCheckUrl>
  <dataCenterInfo class="com.netflix.appinfo.InstanceInfo$DefaultDataCenterInfo">
    <name>MyOwn</name>
  </dataCenterInfo>
</instance>"""


def _register() -> bool:
    if not EUREKA_ENABLED:
        return False
    url = f"{EUREKA_SERVER}/apps/{APP_NAME}"
    try:
        r = requests.post(
            url,
            data=_build_instance_xml(),
            headers={"Content-Type": "application/xml"},
            timeout=5,
        )
        if r.status_code in (200, 204):
            print(f"[EUREKA] Registrado {APP_NAME} como {INSTANCE_ID}")
            return True
        print(f"[EUREKA] Registro falló status={r.status_code} body={r.text[:120]}")
    except Exception as e:
        print(f"[EUREKA] Error registrando: {e}")
    return False


def _heartbeat() -> bool:
    if not EUREKA_ENABLED or not _registered:
        return False
    url = f"{EUREKA_SERVER}/apps/{APP_NAME}/{INSTANCE_ID}"
    try:
        r = requests.put(url, timeout=5)
        if r.status_code in (200, 204):
            return True
        # Si Eureka devuelve 404 (p.ej., tras reiniciarse), intentamos re-registrar
        if r.status_code == 404:
            print("[EUREKA] Heartbeat 404 → re-registering")
            ok = _register()
            return ok
        print(f"[EUREKA] Heartbeat fallo status={r.status_code}")
        return False
    except Exception as e:
        print(f"[EUREKA] Error heartbeat: {e}")
        return False


def _deregister():
    if not EUREKA_ENABLED or not _registered:
        return
    url = f"{EUREKA_SERVER}/apps/{APP_NAME}/{INSTANCE_ID}"
    try:
        requests.delete(url, timeout=5)
        print("[EUREKA] Deregistro enviado")
    except Exception as e:
        print(f"[EUREKA] Error deregistro: {e}")


async def start_eureka_registration():
    global _registered, _heartbeat_task
    if not EUREKA_ENABLED:
        print("[EUREKA] Deshabilitado (EUREKA_ENABLED=0)")
        return
    loop = asyncio.get_event_loop()
    for attempt in range(4):
        ok = await loop.run_in_executor(None, _register)
        if ok:
            _registered = True
            break
        await asyncio.sleep(2 ** attempt)
    if not _registered:
        print("[EUREKA] No se logró registrar (seguirá intentando con heartbeats)")
    async def hb_loop():
        print(f"[EUREKA] Heartbeat cada {HEARTBEAT_INTERVAL}s para {INSTANCE_ID}")
        while True:
            _heartbeat()
            await asyncio.sleep(HEARTBEAT_INTERVAL)
    _heartbeat_task = asyncio.create_task(hb_loop())


async def stop_eureka_registration():
    global _heartbeat_task
    if _heartbeat_task:
        _heartbeat_task.cancel()
        try:
            await _heartbeat_task
        except Exception:
            pass
    loop = asyncio.get_event_loop()
    await loop.run_in_executor(None, _deregister)